src/components/SigningModals/TransactionStepper/TransactionStepper.js

```diff
@@ -0,0 +1,247 @@
+    outerBoundsRef,
+    innerBoundsRef,
+    layout,
+    measuring,
+  } = useStepperLayout()
+
+  const canPerformUpdate = useRef(true)
+
+  const stepsCount = steps.length - 1
+
+  const onStart = useCallback(() => {
+    // Don’t animate on first render
+    if (firstStart) {
+      setFirstStart(false)
+    }
+  }, [firstStart])
+
+  const renderStep = (stepIndex, showDivider) => {
+    const title = steps[stepIndex][0]
+    const { status, hash } = stepState[stepIndex]
+
+    return (
+      <li
+        key={stepIndex}
+        css={`
+          display: flex;
+        `}
+      >
+        <Step
+          title={title}
+          number={stepIndex + 1}
+          status={status}
+          showDivider={showDivider}
+          transactionHash={hash}
+        />
+      </li>
+    )
+  }
+
+  const renderSteps = () => {
+    return steps.map((_, index) => {
+      const showDivider = index < stepsCount
+
+      return renderStep(index, showDivider)
+    })
+  }
+
+  const updateStepStatus = useCallback(
+    status => {
+      if (canPerformUpdate.current) {
+        updateStep(['setStatus', stepperStage, status])
+      }
+    },
+    [stepperStage]
+  )
+
+  const updateHash = useCallback(
+    hash => {
+      if (canPerformUpdate.current) {
+        updateStep(['setHash', stepperStage, hash])
+      }
+    },
+    [stepperStage]
+  )
+
+  const handleSign = useCallback(() => {
+    const signProcess = steps[stepperStage][1]
+
+    // Always start new step in the "Prompting" state
+    updateStepStatus(STEP_PROMPTING)
+
+    // Pass state updates as render props to signProcess
+    signProcess({
+      setStepHash: hash => updateHash(hash),
+      setStepWorking: () => updateStepStatus(STEP_WORKING),
+      setStepError: () => updateStepStatus(STEP_ERROR),
+      setStepSuccess: () => {
+        updateStepStatus(STEP_SUCCESS)
+
+        // Advance to next step or fire complete callback
+        if (stepperStage !== stepsCount && canPerformUpdate.current) {
+          setStepperStage(stepperStage + 1)
+        }
+
+        if (stepperStage === stepsCount && canPerformUpdate.current) {
+          onComplete()
+        }
+      },
+    })
+  }, [
+    steps,
+    stepperStage,
+    updateStepStatus,
+    updateHash,
+    onComplete,
+    stepsCount,
+  ])
+
+  useEffect(handleSign, [stepperStage])
+
+  // Prevent async state updates after unmount
+  useEffect(() => {
+    return () => {
+      canPerformUpdate.current = false
+    }
+  }, [])
+
+  return (
+    <div className={className}>
+      <div
+        ref={outerBoundsRef}
+        css={`
+          display: flex;
+          justify-content: center;
+        `}
+      >
+        <ul
+          ref={innerBoundsRef}
+          css={`
+            padding: 0;
+            display: flex;
+            flex-direction: ${layout === 'single' ? 'column' : 'row'};
+            visibility: ${measuring ? 'hidden' : 'visible'};
+          `}
+        >
+          {layout === 'single' && (
```

Comment:
We may want to consider a special "single step" case, since this is going to be the most frequent number of transactions signed (we generally want to lower the number!).

I think the only thing that might look out of place at the moment is "1 out of 1 transactions" in the small viewport mode.

Reply:
Agreed, for now I'm only going to add the text desc when total count > 1